File: src/my_project/medium_problems/from101to150/reverse_linked_list_ii.py
from typing import List, Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        prev = None
        current = head
        m = left
        n = right
        i = 1
        while i < m and current:
            prev = current
            current = current.next
            i += 1
        start = current
        end = None
        while i <= n and current:
            next = current.next
            current.next = end
            end = current
            current = next
            i += 1

        start.next = current

        if not prev:
            head = end
        else:
            prev.next = end

        return head

File: my_project_typescript/medium_problems_typescript/from51to100/reverseLinkedListII.ts

  class ListNode {
      val: number
      next: ListNode | null
      constructor(val?: number, next?: ListNode | null) {
          this.val = (val===undefined ? 0 : val)
          this.next = (next===undefined ? null : next)
      }
  }


function reverseBetween(head: ListNode | null, left: number, right: number): ListNode | null {
    if (!head) return null;
     if (left === right) return head;
     let current = head, previous = null, counter = 0;
     while (current != null && counter < left - 1) {
       previous = current;
       current = <ListNode>current.next;
       counter++;
     }
     counter = 0;
     const last_node_of_first_part = previous;
     const last_node_of_sub_list = current;
     while (current != null && counter < right - left + 1) {
       let temp = current.next;
       current.next = previous;
       previous = current;
       current = <ListNode>temp;
       counter++;
     }
   
     if (last_node_of_first_part) {
       last_node_of_first_part.next = previous;
     } else {
       head = previous;
     }
     last_node_of_sub_list.next = current;
     return head;
     }